### PR TITLE
Deduplicate packages in the install plan after the solver removes qualifiers.

### LINKED ISSUE #4026: Solver produces lots of duplicate instances in output
I was trying to tighten up the invariants and assertion checking on the InstallPlan, including adding a duplicate check in `Graph.from(Distinct)List` when I discovered that the solver actually produces a lot of duplicates, though all copies are in fact identical.

Here's example output:
```
internal error: could not construct a valid install plan.
The proposed (invalid) plan contained the following problems:
Package Cabal-1.22.5.0 has duplicate instances. There are 11 instances, of which 1 are distinct.
Package array-0.5.1.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package base-4.8.2.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package binary-0.7.5.0 has duplicate instances. There are 11 instances, of which 1 are distinct.
Package bytestring-0.10.6.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package containers-0.5.6.2 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package deepseq-1.4.1.1 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package directory-1.2.2.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package filepath-1.4.0.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package ghc-prim-0.4.0.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package integer-gmp-1.0.0.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package pretty-1.1.2.0 has duplicate instances. There are 12 instances, of which 1 are distinct.
Package process-1.2.3.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package rts-1.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package template-haskell-2.10.0.0 has duplicate instances. There are 12 instances, of which 1 are distinct.
Package time-1.5.0.1 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package transformers-0.4.2.0 has duplicate instances. There are 12 instances, of which 1 are distinct.
Package unix-2.7.1.0 has duplicate instances. There are 14 instances, of which 1 are distinct.
Package hsc2hs-0.68 has duplicate instances. There are 2 instances, of which 1 are distinct.
Package old-locale-1.0.0.7 has duplicate instances. There are 12 instances, of which 1 are distinct.
Package old-time-1.1.0.3 has duplicate instances. There are 12 instances, of which 1 are distinct.
```

For now, I've adjusted the duplicate check to not fail when the duplicates are in fact completely identical. It seems like it's worth checking if this is an indication of a problem and it'd probably be healthier if the solver didn't produce duplicates.

I'll file a PR in a moment with the patches that implement the stronger checking, and it'll be obvious from that what to do to remove the identical instance hack, so that the duplicates get caught as an error.

\cc @kosmikus @grayjay

### PR DESCRIPTION
**Pull request by grayjay**
Fixes #4026

### CODE REVIEW

File: cabal-install/Distribution/Solver/Modular.hs [hunk 1]
```diff
@@ -44,8 +50,12 @@ modularResolver sc (Platform arch os) cinfo iidx sidx pkgConfigDB pprefs pcs pns
...
           pair lpc = (pcName $ unlabelPackageConstraint lpc, [lpc])
 
-      -- Results have to be converted into an install plan.
-      postprocess a rdm = map (convCP iidx sidx) (toCPs a rdm)
+      -- Results have to be converted into an install plan. 'convCP' removes
+      -- package qualifiers, which means that linked packages become duplicates
+      -- and can be removed.
+      -- TODO: Use ordNubBy instead of nubBy.
+      postprocess a rdm = nubBy ((==) `on` nodeKey) $
+                          map (convCP iidx sidx) (toCPs a rdm)
```
Comment (grayjay) [inline cabal-install/Distribution/Solver/Modular.hs]: This can use `ordNubBy` after #4027 is merged.

**Comment by ezyang** [comment]
LGTM!

**Comment by kosmikus** [comment]
LGTM.

**Comment by grayjay** [comment]
Thank you!

**Pull request merged**